File: tools/ffi/python/assetstudio_worker_pool.py
#!/usr/bin/env python3
"""Python sample client for the Rust assetstudio_ffi_worker pool bridge."""

from __future__ import annotations

import argparse
import json
import queue
import struct
import subprocess
import threading
from dataclasses import dataclass
from pathlib import Path
from typing import Any


@dataclass
class WorkerResponse:
    response: dict[str, Any]
    payload: bytes


class AssetStudioWorker:
    def __init__(self, worker_path: str | Path, ffi_library: str | Path):
        self._next_id = 1
        self._lock = threading.Lock()
        worker_path = Path(worker_path).resolve()
        ffi_library = Path(ffi_library).resolve()
        self.proc = subprocess.Popen(
            [
                str(worker_path),
                "--server",
                "--ffi-library",
                str(ffi_library),
            ],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            cwd=Path(ffi_library).resolve().parent,
        )

    def call(self, operation: str, request: dict[str, Any]) -> WorkerResponse:
        with self._lock:
            request_id = self._next_id
            self._next_id += 1
            frame = json.dumps(
                {
                    "id": request_id,
                    "request": {
                        "operation": operation,
                        "request": request,
                    },
                },
                separators=(",", ":"),
            ).encode("utf-8")
            self._write_frame(frame)
            response = json.loads(self._read_frame())
            if response.get("id") != request_id:
                raise RuntimeError(
                    f"worker response id mismatch: expected {request_id}, got {response.get('id')}"
                )
            if response.get("error"):
                raise RuntimeError(response["error"])
            payload_len = int(response.get("payload_len") or 0)
            payload = b""
            if response.get("payload_file"):
                payload_path = Path(response["payload_file"])
                payload = payload_path.read_bytes()
                payload_path.unlink(missing_ok=True)
            elif payload_len:
                payload = self._read_frame()
            if len(payload) != payload_len:
                raise RuntimeError(
                    f"worker payload length mismatch: expected {payload_len}, got {len(payload)}"
                )
            return WorkerResponse(response=response, payload=payload)

    def close(self) -> None:
        if self.proc.stdin:
            try:
                self.proc.stdin.close()
            except BrokenPipeError:
                pass
        try:
            self.proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            self.proc.kill()
            self.proc.wait()

    def _write_frame(self, payload: bytes) -> None:
        if self.proc.stdin is None:
            raise RuntimeError("worker stdin is closed")
        self.proc.stdin.write(struct.pack("<Q", len(payload)))
        self.proc.stdin.write(payload)
        self.proc.stdin.flush()

    def _read_frame(self) -> bytes:
        if self.proc.stdout is None:
            raise RuntimeError("worker stdout is closed")
        header = self.proc.stdout.read(8)
        if len(header) != 8:
            stderr = (
                self.proc.stderr.read().decode("utf-8", errors="replace")
                if self.proc.stderr
                else ""
            )
            raise RuntimeError(f"worker closed stdout: {stderr.strip()}")
        size = struct.unpack("<Q", header)[0]
        payload = self.proc.stdout.read(size)
        if len(payload) != size:
            raise RuntimeError(f"truncated worker frame: expected {size}, got {len(payload)}")
        return payload


class WorkerLease:
    def __init__(self, pool: "AssetStudioWorkerPool", worker: AssetStudioWorker):
        self._pool = pool
        self.worker = worker
        self._released = False

    def __enter__(self) -> AssetStudioWorker:
        return self.worker

    def __exit__(self, exc_type, exc, tb) -> None:
        self.release()

    def release(self) -> None:
        if not self._released:
            self._released = True
            self._pool.release(self.worker)


class AssetStudioWorkerPool:
    def __init__(self, worker_path: str | Path, ffi_library: str | Path, size: int):
        self._workers: queue.LifoQueue[AssetStudioWorker] = queue.LifoQueue()
        self._all_workers = [
            AssetStudioWorker(worker_path, ffi_library) for _ in range(max(1, size))
        ]
        for worker in self._all_workers:
            self._workers.put(worker)

    def acquire(self) -> WorkerLease:
        return WorkerLease(self, self._workers.get())

    def release(self, worker: AssetStudioWorker) -> None:
        self._workers.put(worker)

    def close(self) -> None:
        for worker in self._all_workers:
            worker.close()


def response_body(output: WorkerResponse, expected: str) -> dict[str, Any]:
    outer = output.response.get("response") or {}
    if outer.get("operation") != expected:
        raise RuntimeError(f"unexpected worker response: {outer.get('operation')}, wanted {expected}")
    body = outer.get("response") or {}
    if not body.get("success", False):
        raise RuntimeError(body.get("error") or f"{expected} failed")
    return body


def open_context(worker: AssetStudioWorker, bundle: str, unity_version: str) -> int:
    output = worker.call(
        "context_open",
        {
            "input_path": bundle,
            "asset_types": [],
            "unity_version": unity_version,
            "filter_exclude_mode": False,
            "filter_with_regex": False,
            "filter_by_name": None,
            "filter_by_container": None,
            "filter_by_path_ids": [],
            "load_all_assets": True,
            "include_assets": False,
        },
    )
    return int(response_body(output, "context_open")["context_id"])


def list_all_objects(worker: AssetStudioWorker, context_id: int) -> list[dict[str, Any]]:
    assets: list[dict[str, Any]] = []
    offset = 0
    while True:
        output = worker.call(
            "context_list_objects",
            {"context_id": context_id, "offset": offset, "limit": 2048},
        )
        body = response_body(output, "context_list_objects")
        assets.extend(body.get("assets") or [])
        next_offset = body.get("next_offset")
        if next_offset is None:
            return assets
        offset = int(next_offset)


def read_texture2d(worker: AssetStudioWorker, context_id: int, assets: list[dict[str, Any]]) -> dict[str, Any]:
    textures = [asset for asset in assets if asset.get("type") == "Texture2D"]
    output = worker.call(
        "context_read_objects",
        {
            "context_id": context_id,
            "objects": [
                {"path_id": asset["path_id"], "kind": "image", "image_format": "raw_rgba"}
                for asset in textures
            ],
        },
    )
    body = response_body(output, "context_read_objects")
    return {
        "requested": len(textures),
        "payload_len": len(output.payload),
        "reads": [
            {
                "path_id": read.get("asset", {}).get("path_id"),
                "success": read.get("success"),
                "payload_kind": read.get("payload_kind"),
                "payload_len": read.get("payload_len"),
                "error": read.get("error"),
            }
            for read in body.get("reads", [])
        ],
    }


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--ffi-library", required=True)
    parser.add_argument("--ffi-worker", default="target/release/assetstudio_ffi_worker")
    parser.add_argument("--bundle", required=True)
    parser.add_argument("--unity-version", default="2022.3.21f1")
    parser.add_argument("--pool-size", type=int, default=2)
    parser.add_argument("--read-images", action="store_true")
    args = parser.parse_args()

    pool = AssetStudioWorkerPool(args.ffi_worker, args.ffi_library, args.pool_size)
    try:
        with pool.acquire() as worker:
            context_id = open_context(worker, str(Path(args.bundle).resolve()), args.unity_version)
            try:
                assets = list_all_objects(worker, context_id)
                types: dict[str, int] = {}
                for asset in assets:
                    types[asset.get("type") or ""] = types.get(asset.get("type") or "", 0) + 1
                result: dict[str, Any] = {"asset_count": len(assets), "types": types}
                if args.read_images:
                    result["image_reads"] = read_texture2d(worker, context_id, assets)
                print(json.dumps(result, indent=2, ensure_ascii=False))
            finally:
                worker.call("context_close", {"context_id": context_id})
    finally:
        pool.close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/ffi/node/assetstudio_worker_pool.mjs
#!/usr/bin/env node
// Node.js sample client for the Rust assetstudio_ffi_worker pool bridge.
//
// Zero dependencies: speaks the worker's length-prefixed (u64 LE) JSON frame
// protocol over stdin/stdout, so no native FFI module is needed. Build the
// worker first: cargo build --release --bin assetstudio_ffi_worker

import { spawn } from "node:child_process";
import { once } from "node:events";
import { readFile, unlink } from "node:fs/promises";
import { resolve, dirname } from "node:path";
import { parseArgs } from "node:util";

const MAX_FRAME_SIZE = 256 * 1024 * 1024;

class FrameReader {
  constructor(stream) {
    this.stream = stream;
    this.chunks = [];
    this.length = 0;
  }

  async read() {
    const header = await this.take(8);
    const size = header.readBigUInt64LE(0);
    if (size > BigInt(MAX_FRAME_SIZE)) {
      throw new Error(`worker frame too large: ${size}`);
    }
    return this.take(Number(size));
  }

  async take(size) {
    for (;;) {
      if (this.length >= size) {
        const buffer = Buffer.concat(this.chunks, this.length);
        this.chunks = size < buffer.length ? [buffer.subarray(size)] : [];
        this.length = buffer.length - size;
        return buffer.subarray(0, size);
      }
      const [chunk] = await Promise.race([
        once(this.stream, "data"),
        once(this.stream, "end").then(() => {
          throw new Error("worker closed stdout");
        }),
      ]);
      this.chunks.push(chunk);
      this.length += chunk.length;
    }
  }
}

export class AssetStudioWorker {
  constructor(workerPath, ffiLibrary) {
    workerPath = resolve(workerPath);
    ffiLibrary = resolve(ffiLibrary);
    this.nextId = 1;
    this.queue = Promise.resolve();
    this.proc = spawn(workerPath, ["--server", "--ffi-library", ffiLibrary], {
      cwd: dirname(ffiLibrary),
      stdio: ["pipe", "pipe", "inherit"],
    });
    this.reader = new FrameReader(this.proc.stdout);
  }

  call(operation, request) {
    // Serialize calls per worker: the protocol is strictly request/response.
    const result = this.queue.then(() => this.dispatch(operation, request));
    this.queue = result.catch(() => {});
    return result;
  }

  async dispatch(operation, request) {
    const id = this.nextId;
    this.nextId += 1;
    const frame = Buffer.from(JSON.stringify({ id, request: { operation, request } }), "utf-8");
    const header = Buffer.alloc(8);
    header.writeBigUInt64LE(BigInt(frame.length), 0);
    this.proc.stdin.write(header);
    this.proc.stdin.write(frame);

    const response = JSON.parse((await this.reader.read()).toString("utf-8"));
    if (response.id !== id) {
      throw new Error(`worker response id mismatch: expected ${id}, got ${response.id}`);
    }
    if (response.error) {
      throw new Error(response.error);
    }
    const payloadLen = response.payload_len ?? 0;
    let payload = Buffer.alloc(0);
    if (response.payload_file) {
      // Large payloads are spilled to a temp file; the client owns its cleanup.
      payload = await readFile(response.payload_file);
      await unlink(response.payload_file).catch(() => {});
    } else if (payloadLen > 0) {
      payload = await this.reader.read();
    }
    if (payload.length !== payloadLen) {
      throw new Error(`worker payload length mismatch: expected ${payloadLen}, got ${payload.length}`);
    }
    return { response, payload };
  }

  async close() {
    this.proc.stdin.end();
    if (this.proc.exitCode === null) {
      await once(this.proc, "exit");
    }
  }
}

export class AssetStudioWorkerPool {
  constructor(workerPath, ffiLibrary, size) {
    this.workers = Array.from({ length: Math.max(1, size) }, () => new AssetStudioWorker(workerPath, ffiLibrary));
    this.free = [...this.workers];
    this.waiters = [];
  }

  async acquire() {
    if (this.free.length === 0) {
      await new Promise((resolveWaiter) => this.waiters.push(resolveWaiter));
    }
    return this.free.pop();
  }

  release(worker) {
    this.free.push(worker);
    const waiter = this.waiters.shift();
    if (waiter) {
      waiter();
    }
  }

  async close() {
    await Promise.all(this.workers.map((worker) => worker.close()));
  }
}

function responseBody(output, expected) {
  const outer = output.response.response ?? {};
  if (outer.operation !== expected) {
    throw new Error(`unexpected worker response: ${outer.operation}, wanted ${expected}`);
  }
  const body = outer.response ?? {};
  if (!body.success) {
    throw new Error(body.error ?? `${expected} failed`);
  }
  return body;
}

async function openContext(worker, bundle, unityVersion) {
  const output = await worker.call("context_open", {
    input_path: bundle,
    asset_types: [],
    unity_version: unityVersion,
    filter_exclude_mode: false,
    filter_with_regex: false,
    filter_by_name: null,
    filter_by_container: null,
    filter_by_path_ids: [],
    load_all_assets: true,
    include_assets: false,
  });
  return responseBody(output, "context_open").context_id;
}

async function listAllObjects(worker, contextId) {
  const assets = [];
  let offset = 0;
  for (;;) {
    const output = await worker.call("context_list_objects", { context_id: contextId, offset, limit: 2048 });
    const body = responseBody(output, "context_list_objects");
    assets.push(...(body.assets ?? []));
    if (body.next_offset === null || body.next_offset === undefined) {
      return assets;
    }
    offset = body.next_offset;
  }
}

async function readTexture2D(worker, contextId, assets) {
  const textures = assets.filter((asset) => asset.type === "Texture2D");
  // Upper bound for the packed payload block; compressed GPU formats can expand
  // up to ~16x when decoded to raw RGBA. Above the worker's spill threshold this
  // makes the worker stream payloads through a sparse temp file instead of
  // memory. 0 keeps the in-memory path.
  const payloadCapacityHint = textures.reduce(
    (total, asset) => total + Math.max(0, asset.size ?? 0) * 16 + 1024 * 1024,
    0
  );
  const output = await worker.call("context_read_objects", {
    context_id: contextId,
    objects: textures.map((asset) => ({ path_id: asset.path_id, kind: "image", image_format: "raw_rgba" })),
    payload_capacity_hint: payloadCapacityHint,
  });
  const body = responseBody(output, "context_read_objects");
  return {
    requested: textures.length,
    payload_len: output.payload.length,
    reads: (body.reads ?? []).map((read) => ({
      path_id: read.asset?.path_id,
      success: read.success,
      payload_kind: read.payload_kind,
      payload_len: read.payload_len,
      error: read.error,
    })),
  };
}

async function main() {
  const { values } = parseArgs({
    options: {
      "ffi-library": { type: "string" },
      "ffi-worker": { type: "string", default: "target/release/assetstudio_ffi_worker" },
      bundle: { type: "string" },
      "unity-version": { type: "string", default: "2022.3.21f1" },
      "pool-size": { type: "string", default: "2" },
      "read-images": { type: "boolean", default: false },
    },
  });
  if (!values["ffi-library"] || !values.bundle) {
    console.error(
      "usage: node assetstudio_worker_pool.mjs --ffi-library <path> --bundle <path> [--ffi-worker <path>] [--pool-size <n>] [--read-images]"
    );
    return 2;
  }

  const pool = new AssetStudioWorkerPool(values["ffi-worker"], values["ffi-library"], Number(values["pool-size"]));
  try {
    const worker = await pool.acquire();
    try {
      const contextId = await openContext(worker, resolve(values.bundle), values["unity-version"]);
      try {
        const assets = await listAllObjects(worker, contextId);
        const types = {};
        for (const asset of assets) {
          types[asset.type ?? ""] = (types[asset.type ?? ""] ?? 0) + 1;
        }
        const result = { asset_count: assets.length, types };
        if (values["read-images"]) {
          result.image_reads = await readTexture2D(worker, contextId, assets);
        }
        console.log(JSON.stringify(result, null, 2));
      } finally {
        await worker.call("context_close", { context_id: contextId });
      }
    } finally {
      pool.release(worker);
    }
  } finally {
    await pool.close();
  }
  return 0;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().then(
    (code) => {
      process.exitCode = code;
    },
    (error) => {
      console.error(error);
      process.exitCode = 1;
    }
  );
}
